finality-aleph/src/data_io/data_provider.rs

```diff
@@ -296,3 +309,126 @@ impl<B: BlockT> aleph_bft::DataProvider<AlephData<B>> for DataProvider<B> {
         data
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use crate::{
+        data_io::{
+            data_provider::{ChainTracker, ChainTrackerConfig},
+            AlephData, MAX_DATA_BRANCH_LEN,
+        },
+        testing::{client_chain_builder::ClientChainBuilder, mocks::aleph_data_from_blocks},
+        SessionBoundaries, SessionId, SessionPeriod,
+    };
+    use futures::channel::oneshot;
+    use std::future::Future;
+    use std::{sync::Arc, time::Duration};
+    use substrate_test_runtime_client::{
+        runtime::Block, DefaultTestClientBuilderExt, TestClientBuilder, TestClientBuilderExt,
+    };
+    use tokio::time::sleep;
+
+    const SESSION_LEN: u32 = 100;
+    // The lower the interval the less time the tests take, however setting this too low might cause
+    // the tests to fail. Even though 1ms works with no issues, we set it to 5ms for safety.
+    const REFRESH_INTERVAL_MILLIS: u64 = 5;
+
+    fn prepare_chain_tracker_test() -> (
+        impl Future<Output = ()>,
+        oneshot::Sender<()>,
+        ClientChainBuilder,
+        impl aleph_bft::DataProvider<AlephData<Block>>,
+    ) {
+        let (client, select_chain) = TestClientBuilder::new().build_with_longest_chain();
```

Comment:
why only the non-creator client gets the head-selection policy?

Reply:
In general, the creator client is only used to create new blocks. Only the second -- "main" client we track in terms of block import and finalization, so this one has `SelectChain` that we then use in `DataProvider`. So the answer is, we could also create `select_chain` for the other client, but there is no use for it. Note that `SelectChain` is some kind of a wrapper over `Backend` only (it has read-only access to backend AFAIK), so it does not affect block creation, import, finalization, nor what the `best_block` is.